x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NativeMemoryCapacity.java

```diff
@@ -48,61 +69,151 @@ public NativeMemoryCapacity(long tierMlNativeMemoryRequirement, long nodeMlNativ
+        // It shouldn't be seen very often.
+        long tierMlNativeMemoryRequirementPerZoneExcludingOverhead = (numMlAvailabilityZones > 1)
+            ? ((tierMlNativeMemoryRequirementExcludingOverhead + numMlAvailabilityZones - 1) / numMlAvailabilityZones)
+            : tierMlNativeMemoryRequirementExcludingOverhead;
+        int numNodesPerZone = (int) ((tierMlNativeMemoryRequirementPerZoneExcludingOverhead
+            + mlNativeMemoryForLargestMlNodeExcludingOverhead - 1) / mlNativeMemoryForLargestMlNodeExcludingOverhead);
+        long tierBasedMlNativeMemoryPerNodeExcludingOverhead = (tierMlNativeMemoryRequirementPerZoneExcludingOverhead + numNodesPerZone - 1)
+            / numNodesPerZone;
+
         // We first need to calculate the required node size given the required native ML memory size.
-        // This way we can accurately determine the required node size AND what the overall memory percentage will be
         long requiredNodeSize = NativeMemoryCalculator.calculateApproxNecessaryNodeSize(
-            nodeMlNativeMemoryRequirement,
+            nodeMlNativeMemoryRequirementExcludingOverhead + NATIVE_EXECUTABLE_CODE_OVERHEAD.getBytes(),
             jvmSize,
             maxMemoryPercent,
             useAuto
         );
-        // We make the assumption that the JVM size is the same across the entire tier
-        // This simplifies calculating the tier as it means that each node in the tier
-        // will have the same dynamic memory calculation. And thus the tier is simply the sum of the memory necessary
-        // times that scaling factor.
-        //
-        // Since this is a _minimum_ node size, the memory percent calculated here is not
-        // necessarily what "auto" will imply after scaling. Because the JVM occupies a
-        // smaller proportion of memory the bigger the node, the memory percent might be
-        // higher than we calculate here, potentially resulting in a bigger ML tier than
-        // required. The effect is most pronounced when the minimum node size is small and
-        // the minimum tier size is a lot bigger. For example, if the minimum node size is
-        // 1GB and the total native ML memory requirement for the tier is 32GB then the memory
-        // percent will be 41%, implying a minimum ML tier size of 78GB. But in reality a
-        // single 64GB ML node would have an auto memory percent of 90%, and 90% of 64GB is
-        // plenty big enough for the 32GB of ML native memory required.
-        // TODO: improve this in the next refactoring
-        double memoryPercentForMl = NativeMemoryCalculator.modelMemoryPercent(requiredNodeSize, jvmSize, maxMemoryPercent, useAuto);
-        double inverseScale = memoryPercentForMl <= 0 ? 0 : 100.0 / memoryPercentForMl;
-        long requiredTierSize = Math.round(Math.ceil(tierMlNativeMemoryRequirement * inverseScale));
```

Comment:
The new code in this method doesn't use floating point arithmetic. I found it caused unpredictable rounding problems because multi-gigabyte byte values we're working with are too close to the maximum 15 digit precision of `double`.  It's safer to stick to integer arithmetic.